### PR TITLE
feat(react-grid): allow showing row summaries

### LINKED ISSUE #1024: How to add Group Summary Row
- [x] I have searched this repository's [issues](https://github.com/devexpress/devextreme-reactive/issues) and believe that this is not a duplicate.

I'd like to be able to add a summary row when column is grouped. Like the example in this picture.
screen shot 2018-05-01 at 5 24 30 pm

### PR DESCRIPTION
**Pull request by kvet**
closes #1024

### CODE REVIEW

File: packages/dx-grid-core/src/plugins/integrated-summary/computeds.js [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+  min: (rows, getValue) => (rows.length
+    ? rows.reduce((acc, row) => Math.min(acc, getValue(row)), Infinity)
+    : null),
+  avg: (rows, getValue) => (rows.length
+    ? rows.reduce((acc, row) => acc + getValue(row), 0) / rows.length
+    : null),
+};
+
+export const defaultSummaryCalculator = (type, rows, getValue) => {
+  const summaryCulculator = defaultSummaryCalculators[type];
```
Comment (SergeyAlexeev) [inline packages/dx-grid-core/src/plugins/integrated-summary/computeds.js]: Culculator -> Calculator

File: packages/dx-grid-core/src/plugins/table-summary-row/computeds.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+export const tableRowsWithTotalSummaries = footerRows => [
+  { key: TABLE_TOTAL_SUMMARY_TYPE, type: TABLE_TOTAL_SUMMARY_TYPE },
+  ...footerRows,
+];
+
+export const tableRowsWithSummaries = (tableRows, getRowLevelKey, isGroupRow, getRowId) => {
+  if (!getRowLevelKey) return tableRows;
+
+  const result = [];
+  const closeLevel = (level) => {
```
Comment (MaximKudriavtsev) [inline packages/dx-grid-core/src/plugins/table-summary-row/computeds.js]: Can we extract this function?

File: packages/dx-grid-core/src/plugins/table-summary-row/helpers.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+export const isTotalSummaryTableRow = tableRow => tableRow.type === TABLE_TOTAL_SUMMARY_TYPE;
+export const isGroupSummaryTableRow = tableRow => tableRow.type === TABLE_GROUP_SUMMARY_TYPE;
+export const isTreeSummaryTableRow = tableRow => tableRow.type === TABLE_TREE_SUMMARY_TYPE;
+
+export const getColumnSummaries = (summaryItems, columnName, summaryValues) => {
+  const columnSummaryItems = summaryItems
+    .map((item, index) => [item, index])
+    .filter(([item]) => item.columnName === columnName);
+  return columnSummaryItems
+    .map(([item, index]) => ({ type: item.type, value: summaryValues[index] }));
```
Comment (MaximKudriavtsev) [inline packages/dx-grid-core/src/plugins/table-summary-row/helpers.js]: Can we remove `return` from this function? I mean we can use `return summaryItems.map().filter().map()`

File: packages/dx-grid-core/src/plugins/table-summary-row/helpers.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  tableRow, tableColumn,
+) => tableRow.type === TABLE_GROUP_SUMMARY_TYPE && tableColumn.type === TABLE_DATA_TYPE;
+export const isTreeSummaryTableCell = (
+  tableRow, tableColumn,
+) => tableRow.type === TABLE_TREE_SUMMARY_TYPE && tableColumn.type === TABLE_DATA_TYPE;
+export const isTotalSummaryTableRow = tableRow => tableRow.type === TABLE_TOTAL_SUMMARY_TYPE;
+export const isGroupSummaryTableRow = tableRow => tableRow.type === TABLE_GROUP_SUMMARY_TYPE;
+export const isTreeSummaryTableRow = tableRow => tableRow.type === TABLE_TREE_SUMMARY_TYPE;
+
+export const getColumnSummaries = (summaryItems, columnName, summaryValues) => {
```
Comment (MaximKudriavtsev) [inline packages/dx-grid-core/src/plugins/table-summary-row/helpers.js]: This function doesn't have tests.

File: packages/dx-react-grid/docs/guides/summary-row.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
+# React Grid - Sorting
```
Comment (MaximKudriavtsev) [inline packages/dx-react-grid/docs/guides/summary-row.md]: Sorting -> Summary Row

File: packages/dx-grid-core/src/plugins/table-summary-row/computeds.js (hunk 2 above)
Comment (kvet) [reply packages/dx-grid-core/src/plugins/table-summary-row/computeds.js]: No, it have closures to local variables

File: packages/dx-grid-core/src/plugins/integrated-summary/computeds.js [hunk 6]
```diff
@@ -0,0 +1,146 @@
+const defaultSummaryCalculators = {
+  count: rows => rows.length,
+  sum: (rows, getValue) => rows.reduce((acc, row) => acc + getValue(row), 0),
+  max: (rows, getValue) => (rows.length
+    ? rows.reduce((acc, row) => Math.max(acc, getValue(row)), -Infinity)
```
[the remark this answers is not in the document]
Comment (kvet) [reply packages/dx-grid-core/src/plugins/integrated-summary/computeds.js]: No, I can't. We should map rows to cell values

File: packages/dx-grid-core/src/plugins/integrated-summary/computeds.js [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+  summaryItems,
+  getCellValue,
+  getRowLevelKey,
+  isGroupRow,
+  getCollapsedRows,
+  calculator = defaultSummaryCalculator,
+) => {
+  const plainRows = rows.reduce((acc, row) => {
+    if (getRowLevelKey && getRowLevelKey(row)) {
+      if (!isGroupRow || !isGroupRow(row)) {
```
[the remark this answers is not in the document]
Comment (kvet) [reply packages/dx-grid-core/src/plugins/integrated-summary/computeds.js]: I need to use negation here. So, I think that expression with opened parentheses is more readable

File: packages/dx-react-grid/docs/reference/custom-summary.md [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+- [SummaryState](summary-state.md)
+
+### Properties
+
+Name | Type | Default | Description
+-----|------|---------|------------
+totalValues | Array&lt;any&gt; | Summary values related to all rows.
+groupValues | { [key: [GroupKey](grouping-state.md#groupkey)]: Array&lt;any&gt; } | Summary values related to group rows.
+treeValues | { [key: number &#124; string]: Array&lt;any&gt; } | Summary values related to rows that contains children rows.
```
Comment (MaximKudriavtsev) [inline packages/dx-react-grid/docs/reference/custom-summary.md]: `Description` section stay in the `Default`
image

File: packages/dx-react-grid/docs/reference/table-summary-row.md [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+
+## Localization Messages
+
+Field | Type | Default | Description
+------|------|---------|------------
+count? | string | 'Count' | Specifies the 'count' summary text.
+sum? | string | 'Sum' | Specifies the 'sum' summary text.
+min? | string | 'Min' | Specifies the 'min' summary text.
+max? | string | 'Max' | Specifies the 'max' summary text.
+avg? | string | 'Avg' | Specifies the 'avg' summary text.
```
Comment (MaximKudriavtsev) [inline packages/dx-react-grid/docs/reference/table-summary-row.md]: Maybe in the description, we should explain more?
For example, `Specifies the text for average type`

File: packages/dx-react-grid/docs/reference/table-summary-row.md [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+totalRowComponent | ComponentType&lt;[Table.RowProps](table.md#tablerowprops)&gt; | A component that renders a row with total summaries.
+TableSummaryRow.GroupRow | ComponentType&lt;[Table.RowProps](table.md#tablerowprops)&gt; | A component that renders a row with group summaries.
+TableSummaryRow.TreeRow | ComponentType&lt;[Table.RowProps](table.md#tablerowprops)&gt; | A component that renders a row with tree summaries.
+TableSummaryRow.TotalCell | ComponentType&lt;[TableSummaryRow.CellProps](#tablesummaryrowcellprops)&gt; | A component that renders a cell with total summaries.
+TableSummaryRow.GroupCell | ComponentType&lt;[TableSummaryRow.CellProps](#tablesummaryrowcellprops)&gt; | A component that renders a cell with group summaries.
+TableSummaryRow.TreeCell | ComponentType&lt;[TableSummaryRow.CellProps](#tablesummaryrowcellprops)&gt; | A component that renders a cell with tree summaries.
+TableSummaryRow.TreeColumnCell | ComponentType&lt;[TableSummaryRow.CellProps](#tablesummaryrowcellprops)&gt; | A component that renders a tree cell within a tree summary row.
```
Comment (MaximKudriavtsev) [inline packages/dx-react-grid/docs/reference/table-summary-row.md]: Is not clear to me. What is a difference between TreeCell and Tree Column Cell?

File: packages/dx-react-grid/docs/guides/summary-row.md [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+The `CustomSummary` plugin allows providing data summary values calculated externally. It may be received from the server or computed locally.
+
+In the following example, customly calculated total summary row is shown. Pass summary items array to the `totalItems` property of the `SummaryState` plugin and summary values to the `totalValues` property of the `CustomSummary` plugin.
+
+.embedded-demo({ "path": "grid-summary/custom-summary", "showThemeSelector": true })
+
+The same is possible with grouped and tree data. Pass the `groupValues` and `treeValues` data object to the `CustomSummary` plugin.
+
+## Custom Algorithm
```
Comment (MaximKudriavtsev) [inline packages/dx-react-grid/docs/guides/summary-row.md]: Just Idea. Maybe we should add to this guide `Appearance Customization` section and create a demo with summary rows with other color?

File: packages/dx-grid-core/src/plugins/integrated-summary/computeds.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (alexey-semikozov) [reply packages/dx-grid-core/src/plugins/integrated-summary/computeds.js]: I mean
```
Math.min(...rows.map(getValue), -Infinity)
```
But maybe better to keep it that way.

File: packages/dx-grid-core/src/plugins/integrated-summary/computeds.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (alexey-semikozov) [reply packages/dx-grid-core/src/plugins/integrated-summary/computeds.js]: Oh, it breaks my mind. If you think it's ok, keep it that way

File: packages/dx-grid-core/src/plugins/integrated-summary/computeds.js [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+  }, {});
+  levels.forEach((level) => {
+    summaries[level.row.compoundKey] = rowsSummary(
+      level.rows, summaryItems, getCellValue, calculator,
+    );
+  });
+  return summaries;
+};
+
+export const treeSummaryValues = (
```
[the remark this answers is not in the document]
Comment (kvet) [reply packages/dx-grid-core/src/plugins/integrated-summary/computeds.js]: I agree that there are same code in this functions but they are not the same. Read them attentively. There are too many differences. I don't sure that I can extract function.

File: packages/dx-react-grid/docs/reference/integrated-summary.md [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+
+### Dependencies
+
+- [SummaryState](summary-state.md)
+
+### Properties
+
+Name | Type | Default | Description
+-----|------|---------|------------
+calculator | (type: [SummaryType](summary-state.md#summarytype), rows: Array&lt;any&gt;, getValue: (row: any) => any) => any | | A summary calculator.
```
Comment (gsobolev) [inline packages/dx-react-grid/docs/reference/integrated-summary.md]: What the type of the `getValue` parameter?

File: packages/dx-react-grid/docs/guides/summary-row.md [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+
+## Custom Algorithms
+
+The `IntegratedSummary` plugin allows you to extend the built-in summary types (`count`, `sum`, `max`, `min`, `avg`) with custom types. In the followith example, a custom `median` summary type is defined:
+
+.embedded-demo({ "path": "grid-summary/custom-algorithm", "showThemeSelector": true })
+
+## Avoiding Summary Formatting
+
+If a `DataTypeProvider` plugin is assigned to a column, its value formatting component is applied to a summary row. Use the `formattlessSummaryTypes` property to specify summary types that should not be formatted. The following example shows the custom `overprice` summary type that is not formatted:
```
Comment (dirk-pieterse) [inline packages/dx-react-grid/docs/guides/summary-row.md]: the custom -> perhaps 'a' custom?

File: packages/dx-react-grid/docs/guides/summary-row.md [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+
+The following example demonstrates customly calculated total summary. Pass summary items array to the `SummaryState` plugin's `totalItems` property and summary values to the `totalValues` property.
+
+.embedded-demo({ "path": "grid-summary/custom-summary", "showThemeSelector": true })
+
+Calculate custom group and tree summary in the same manner. Pass the `groupValues` and `treeValues` data objects to the `CustomSummary` plugin.
+
+## Custom Algorithms
+
+The `IntegratedSummary` plugin allows you to extend the built-in summary types (`count`, `sum`, `max`, `min`, `avg`) with custom types. In the followith example, a custom `median` summary type is defined:
```
Comment (dirk-pieterse) [inline packages/dx-react-grid/docs/guides/summary-row.md]: "In the followith example, a custom `median` summary type is defined:" 
followith? 
a custom `median` summary type is defined?

File: packages/dx-react-grid/docs/guides/summary-row.md [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+
+The following example shows tree summary rows. Pass the summary items' array to the `SummaryState` plugin's `treeItems` property.
+
+.embedded-demo({ "path": "grid-summary/tree-summary", "showThemeSelector": true })
+
+### Custom Data Summaries Calculation
+
+The `CustomSummary` plugin allows you to calculate a data summary on a remote server or local machine.
+
+The following example demonstrates customly calculated total summary. Pass summary items array to the `SummaryState` plugin's `totalItems` property and summary values to the `totalValues` property.
```
Comment (dirk-pieterse) [inline packages/dx-react-grid/docs/guides/summary-row.md]: customly?

All these sentences are strange: 'The following example this. Pass this.' Pass this to do what?

File: packages/dx-react-grid/docs/guides/summary-row.md [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+
+The Grid component can calculate a summary for all rows (total summary), row groups (group summary), and tree nodes that contain child nodes (tree summary). You can use the corresponding plugins to calculate the summary and manage its state.
+
+## Related Plugins
+
+The following plugins implement summary features:
+
+- [SummaryState](../reference/summary-state.md) - manages the data summary state
+- [IntegratedSummary](../reference/integrated-summary.md) - performs the built-in data summary calculation
+- [CustomSummary](../reference/custom-summary.md) - provides a capability to calculate a custom data summary
```
Comment (dirk-pieterse) [inline packages/dx-react-grid/docs/guides/summary-row.md]: provides a capability? a capability?

File: packages/dx-react-grid/docs/reference/custom-summary.md [hunk 18]
```diff
@@ -0,0 +1,39 @@
+# CustomSummary Plugin Reference
+
+A plugin that provides a capability to calculate custom summary.
```
Comment (dirk-pieterse) [inline packages/dx-react-grid/docs/reference/custom-summary.md]: Again, "provides a capability"?

Comment (dirk-pieterse) [reply packages/dx-react-grid/docs/reference/custom-summary.md]: and no articles: **a** custom summary

File: packages/dx-react-grid/docs/reference/custom-summary.md [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+
+## Importing
+
+Use the following import statement:
```
Comment (dirk-pieterse) [inline packages/dx-react-grid/docs/reference/custom-summary.md]: Importing? 

"Use the following import statement:" To do what?

File: packages/dx-react-grid/docs/reference/table-summary-row.md [hunk 20]
```diff
@@ -0,0 +1,125 @@
+# TableSummaryRow Plugin Reference
+
+A plugin that renders table rows that display total, group, and tree summary.
```
Comment (dirk-pieterse) [inline packages/dx-react-grid/docs/reference/table-summary-row.md]: A plugin that renders table rows that display total, group, and tree summary -> A plugin that renders table rows that display **a** total, group, and tree summary.

File: packages/dx-react-grid/docs/reference/integrated-summary.md [hunk 21]
```diff
@@ -0,0 +1,48 @@
+# IntegratedSummary Plugin Reference
+
+A plugin that performs built-in data summary calculation.
```
Comment (dirk-pieterse) [inline packages/dx-react-grid/docs/reference/integrated-summary.md]: A plugin that performs built-in data summary calculation -> A plugin that performs **a** built-in data summary calculation.

File: packages/dx-react-grid/docs/reference/table-tree-column.md [hunk 22]
```diff
@@ -104,7 +104,7 @@ TableTreeColumn.Indent | [TableTreeColumn.IndentProps](#tabletreecolumnindentpro
 TableTreeColumn.ExpandButton | [TableTreeColumn.ExpandButtonProps](#tabletreecolumnexpandbuttonprops) | A component that renders a button used to controls a row's expanded state.
 TableTreeColumn.Checkbox | [TableTreeColumn.CheckboxProps](#tabletreecolumncheckboxprops) | A component that renders a checkbox used to control selection.
 
-The additional properties are added to the component's root element.
```
Comment (dirk-pieterse) [inline packages/dx-react-grid/docs/reference/table-tree-column.md]: Specifies whether to render selection controls. Requires the [SelectionState] dependency.
Specifies whether to render Select All checkbox. Requires the [IntegratedSelection] dependency.

Something cannot require a dependency. A dependency means it is a requirement. Rewrite.

File: packages/dx-react-grid/docs/reference/table.md [hunk 23]
```diff
@@ -33,6 +33,7 @@ columnExtensions? | Array&lt;[Table.ColumnExtension](#tablecolumnextension)&gt;
 tableComponent | ComponentType&lt;object&gt; | | A component that renders a table.
```
Comment (dirk-pieterse) [inline packages/dx-react-grid/docs/reference/table.md]: "A plugin that renders Grid data as a table. It contains the Table Row and Table Cell components that can be extended by other plugins **and provides ways** to customize table rows and columns." What provides ways?

File: packages/dx-react-grid/docs/reference/table-tree-column.md (hunk 22 above)
Comment (viterobk) [reply packages/dx-react-grid/docs/reference/table-tree-column.md]: Requires the [...] plugin.

**Pull request merged**